File: algorithms/leetcode/023-merge-k-sorted-lists/merge-k-sorted-lists.js
const mergeKLists = (lists) => {
    if (!lists.length) {
        return null;
    }

    if (lists.length === 1) {
        return lists[0];
    }

    const mergeTwoLists = (l1, l2) => {
        let root = new ListNode(0);
        let node = root;

        while (l1 && l2) {
            if (l1.val < l2.val) {
                node = node.next = l1;
                l1 = l1.next;
            } else {
                node = node.next = l2;
                l2 = l2.next;
            }
        }

        node.next = l1 ? l1 : l2;

        return root.next;
    };

    let list = lists[0];

    for (let i = 1; i < lists.length; i++) {
        list = mergeTwoLists(list, lists[i]);
    }

    return list ? list : [];
};

File: leetcode/023-merge-k-sorted-lists/MergeKSortedLists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists.length == 0) {
            return null;
        }
        ListNode result = lists[0];
        for (int i = 1; i < lists.length; i++) {
            result = mergeTwoLists(result, lists[i]);
        }
        return result;
    }
    
    private ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode head = new ListNode(0);
        ListNode tail = head;
        
        while (list1 != null || list2 != null) {
            if (list2 == null) {
                tail.next = list1;
                list1 = list1.next;
            } else if (list1 == null) {
                tail.next = list2;
                list2 = list2.next;
            } else if (list1.val <= list2.val) {
                tail.next = list1;
                list1 = list1.next;
            } else {
                tail.next = list2;
                list2 = list2.next;
            }
            tail = tail.next;
        }
        
        return head.next;
    }
}
